File: src/GafferOSLUI/OSLImageUI.cpp
//////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2019, Image Engine Design Inc. All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are
//  met:
//
//      * Redistributions of source code must retain the above
//        copyright notice, this list of conditions and the following
//        disclaimer.
//
//      * Redistributions in binary form must reproduce the above
//        copyright notice, this list of conditions and the following
//        disclaimer in the documentation and/or other materials provided with
//        the distribution.
//
//      * Neither the name of John Haddon nor the names of
//        any other contributors to this software may be used to endorse or
//        promote products derived from this software without specific prior
//        written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//////////////////////////////////////////////////////////////////////////


#include "GafferOSL/OSLImage.h"
#include "GafferOSL/ClosurePlug.h"

#include "GafferUI/Nodule.h"
#include "GafferUI/NoduleLayout.h"
#include "GafferUI/PlugAdder.h"

#include "Gaffer/Metadata.h"
#include "Gaffer/MetadataAlgo.h"
#include "Gaffer/ScriptNode.h"
#include "Gaffer/UndoScope.h"
#include "Gaffer/NameValuePlug.h"
#include "Gaffer/PlugAlgo.h"

#include "IECore/CompoundData.h"

#include "boost/algorithm/string.hpp"
#include "boost/bind/bind.hpp"

using namespace std;
using namespace boost::placeholders;
using namespace Gaffer;
using namespace GafferUI;

namespace
{

std::string cleanupChannelName( std::string s )
{
	boost::erase_all( s, "RGBA" );
	boost::erase_all( s, "RGB" );
	return s;
}

class OSLImagePlugAdder : public PlugAdder
{

	public :

		OSLImagePlugAdder( GraphComponentPtr plugsParent )
			:   m_plugsParent( IECore::runTimeCast<Plug>( plugsParent ) )
		{
			if( ! m_plugsParent )
			{
				throw IECore::Exception( "OSLImageUI::PlugAdder constructor must be passed plug" );
			}
			buttonReleaseSignal().connect( boost::bind( &OSLImagePlugAdder::buttonRelease, this, ::_2 ) );
		}

	protected :

		bool canCreateConnection( const Plug *endpoint ) const override
		{
			IECore::ConstCompoundDataPtr plugAdderOptions = Metadata::value<IECore::CompoundData>( m_plugsParent->node(), "plugAdderOptions" );
			return !availableChannels( plugAdderOptions.get(), endpoint ).empty();
		}

		void createConnection( Plug *endpoint ) override
		{
			IECore::ConstCompoundDataPtr plugAdderOptions = Metadata::value<IECore::CompoundData>( m_plugsParent->node(), "plugAdderOptions" );
			vector<std::string> names = availableChannels( plugAdderOptions.get(), endpoint );

			std::string picked = menuSignal()( "Connect To", names );
			if( !picked.size() )
			{
				return;
			}

			NameValuePlug *newPlug = addPlug( cleanupChannelName( picked ), plugAdderOptions->member<IECore::Data>( picked ) );
			newPlug->valuePlug()->setInput( endpoint );
		}

	private :
		std::set<std::string> usedNames() const
		{
			std::set<std::string> used;
			for( NameValuePlug::Iterator it( m_plugsParent.get() ); !it.done(); ++it )
			{
				// TODO - this method for checking if a plug variesWithContext should probably live in PlugAlgo
				// ( it's based on Switch::variesWithContext )
				PlugPtr sourcePlug = (*it)->namePlug()->source<Gaffer::Plug>();
				bool variesWithContext = sourcePlug->direction() == Plug::Out && IECore::runTimeCast<const ComputeNode>( sourcePlug->node() );
				if( !variesWithContext )
				{
					used.insert( (*it)->namePlug()->getValue() );
				}
			}
			return used;
		}

		NameValuePlug* addPlug( std::string channelName, const IECore::Data *defaultData )
		{
			std::set<std::string> used = usedNames();

			std::string plugName = "channel";
			PlugPtr valuePlug;
			FloatPlugPtr alphaValuePlug;
			if( defaultData )
			{
				const IECore::Color4fData *color4fDefaultData = IECore::runTimeCast<const IECore::Color4fData>( defaultData );
				if( color4fDefaultData )
				{
					const Imath::Color4f &default4 = color4fDefaultData->readable();
					alphaValuePlug = new FloatPlug( "value", Plug::In, default4[3], Imath::limits<float>::min(), Imath::limits<float>::max(), Plug::Flags::Default | Plug::Flags::Dynamic );
					defaultData = new IECore::Color3fData( Imath::Color3f( default4[0], default4[1], default4[2] ) );
				}

				if( used.find( channelName ) != used.end() )
				{
					std::string newName;
					for( int i = 2; ; i++ )
					{
						newName = channelName + std::to_string( i );
						if( used.find( newName ) == used.end() )
						{
							break;
						}
					}
					channelName = newName;
				}
				valuePlug = PlugAlgo::createPlugFromData( "value", Plug::In, Plug::Flags::Default | Plug::Flags::Dynamic, defaultData );
			}
			else
			{
				valuePlug = new GafferOSL::ClosurePlug( "value", Plug::In, Plug::Flags::Default | Plug::Flags::Dynamic );
				plugName = "closure";
				channelName = "";
			}

			UndoScope undoScope( m_plugsParent->ancestor<ScriptNode>() );

			NameValuePlugPtr created = new Gaffer::NameValuePlug( channelName, valuePlug, true, plugName );
			m_plugsParent->addChild( created );
			if( alphaValuePlug )
			{
				std::string alphaChannelName = "A";
				if( channelName.size() )
				{
					alphaChannelName = channelName + ".A";
				}
				m_plugsParent->addChild( new Gaffer::NameValuePlug( alphaChannelName, alphaValuePlug, true, plugName ) );
			}
			return created.get();
		}

		bool buttonRelease( const ButtonEvent &event )
		{
			IECore::ConstCompoundDataPtr plugAdderOptions = Metadata::value<IECore::CompoundData>( m_plugsParent->node(), "plugAdderOptions" );
			vector<std::string> origNames = availableChannels( plugAdderOptions.get() );
			map<std::string, std::string> nameMapping;
			vector<std::string> standardMenuNames;
			vector<std::string> customMenuNames;
			vector<std::string> advancedMenuNames;
			for( auto &n : origNames )
			{
				std::string menuName;
				if( n.substr( 0, 6 ) == "custom" )
				{
					menuName = "Custom/" + n.substr( 6 );
					customMenuNames.push_back( menuName );
				}
				else if( n == "closure" )
				{
					menuName = "Advanced/Closure";
					advancedMenuNames.push_back( menuName );
				}
				else
				{
					menuName = "Standard/" + n;
					standardMenuNames.push_back( menuName );
				}
				nameMapping[ menuName ] = n;
			}

			vector<std::string> menuNames;
			menuNames.insert( menuNames.end(), standardMenuNames.begin(), standardMenuNames.end() );
			menuNames.insert( menuNames.end(), customMenuNames.begin(), customMenuNames.end() );
			menuNames.insert( menuNames.end(), advancedMenuNames.begin(), advancedMenuNames.end() );
			std::string picked = menuSignal()( "Add Input", menuNames );
			if( !picked.size() )
			{
				return false;
			}

			std::string origName = nameMapping[picked];
			addPlug( cleanupChannelName( origName ), plugAdderOptions->member<IECore::Data>(origName) );

			return true;
		}

		// Which channels are available that haven't already been used, and that match the input plug if provided
		vector<std::string> availableChannels( const IECore::CompoundData* plugAdderOptions, const Plug *input = nullptr ) const
		{
			if( !plugAdderOptions )
			{
				throw IECore::Exception( "OSLImageUI::PlugAdder requires plugAdderOptions metadata" );
			}

			IECore::DataPtr matchingDataType;
			const ValuePlug *valueInput = IECore::runTimeCast< const ValuePlug >( input );
			if( valueInput )
			{
				try
				{
					matchingDataType = PlugAlgo::getValueAsData( valueInput );
				}
				catch( ... )
				{
					// If we can't extract data, then it doesn't match any of our accepted plug types
				}
			}

			vector<std::string> result;
			std::set<std::string> used = usedNames();
			for( auto it=plugAdderOptions->readable().begin(); it!=plugAdderOptions->readable().end(); it++ )
			{
				std::string bareLabel = cleanupChannelName( it->first );

				// For plugs that aren't closures or custom, we need to check if we've already
				// used the name
				if( it->second && bareLabel.substr( 0, 6 ) != "custom" && used.find( bareLabel ) != used.end() )
				{
					// Already added
					continue;
				}

				if( input )
				{
					if( input->typeId() == GafferOSL::ClosurePlug::staticTypeId() )
					{
						if( it->second )
						{
							continue;
						}
					}
					else
					{
						if( !matchingDataType || !it->second || matchingDataType->typeId() != it->second->typeId() )
						{
							continue;
						}
					}
				}

				result.push_back( it->first );
			}

			std::sort( result.begin(), result.end() );
			vector<std::string> customSortResult;
			for( const std::string &i : { "RGB", "RGBA", "R", "G", "B", "A" } )
			{
				if( std::find( result.begin(), result.end(), i ) != result.end() )
				{
					customSortResult.push_back( i );
				}
			}
			for( const std::string &i : result )
			{
				if( std::find( customSortResult.begin(), customSortResult.end(), i ) == customSortResult.end() )
				{
					customSortResult.push_back( i );
				}
			}

			return customSortResult;
		}

		PlugPtr m_plugsParent;
};

struct Registration
{
		Registration()
		{
			NoduleLayout::registerCustomGadget( "GafferOSLUI.OSLImageUI.PlugAdder", &create );
		}

	private :

		static GadgetPtr create( GraphComponentPtr parent )
		{
			return new OSLImagePlugAdder( parent );
		}
};

Registration g_registration;

} // namespace

File: python/GafferOSLUI/OSLImageUI.py
##########################################################################
#
#  Copyright (c) 2013-2014, John Haddon. All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#      * Redistributions of source code must retain the above
#        copyright notice, this list of conditions and the following
#        disclaimer.
#
#      * Redistributions in binary form must reproduce the above
#        copyright notice, this list of conditions and the following
#        disclaimer in the documentation and/or other materials provided with
#        the distribution.
#
#      * Neither the name of John Haddon nor the names of
#        any other contributors to this software may be used to endorse or
#        promote products derived from this software without specific prior
#        written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
##########################################################################

import IECore
import Gaffer
import GafferUI

import GafferOSL

import imath
import functools

_channelNamesOptions = {
	"RGB" : IECore.Color3fData( imath.Color3f( 1 ) ),
	"RGBA" : IECore.Color4fData( imath.Color4f( 1 ) ),
	"R" : IECore.FloatData( 1 ),
	"G" : IECore.FloatData( 1 ),
	"B" : IECore.FloatData( 1 ),
	"A" : IECore.FloatData( 1 ),
	"customChannel" : IECore.FloatData( 1 ),
	"customLayer" : IECore.Color3fData( imath.Color3f( 1 ) ),
	"customLayerRGBA" : IECore.Color4fData( imath.Color4f( 1 ) ),
	"closure" : None,
}

##########################################################################
# _ChannelsFooter
##########################################################################

class _ChannelsFooter( GafferUI.PlugValueWidget ) :

	def __init__( self, plug ) :

		row = GafferUI.ListContainer( GafferUI.ListContainer.Orientation.Horizontal )

		GafferUI.PlugValueWidget.__init__( self, row, plug )

		with row :

				GafferUI.Spacer( imath.V2i( GafferUI.PlugWidget.labelWidth(), 1 ) )

				self.__menuButton = GafferUI.MenuButton(
					image = "plus.png",
					hasFrame = False,
					menu = GafferUI.Menu(
						Gaffer.WeakMethod( self.__menuDefinition ),
						title = "Add Input"
					),
					toolTip = "Add Input"
				)

				GafferUI.Spacer( imath.V2i( 1 ), imath.V2i( 999999, 1 ), parenting = { "expand" : True } )

	def _updateFromEditable( self ) :

		self.__menuButton.setEnabled( self._editable() )

	def __menuDefinition( self ) :

		result = IECore.MenuDefinition()
		usedNames = set()
		for p in self.getPlug().children():
			if not Gaffer.PlugAlgo.dependsOnCompute( p ) :
				usedNames.add( p["name"].getValue() )

		# Use a fixed order for some standard options that we want to list in a specific order
		sortedOptions = []
		for label in ["RGB", "RGBA", "R", "G", "B", "A" ]:
			sortedOptions.append( (label, _channelNamesOptions[label] ) )

		for label, defaultData in sorted( _channelNamesOptions.items() ):
			if not label in [ i[0] for i in sortedOptions ]:
				sortedOptions.append( (label, defaultData) )

		categories = { "Standard" : [], "Custom" : [], "Advanced" : [] }
		for label, defaultData in sortedOptions:
			if label == "closure":
				categories["Advanced"].append( ( label, label, defaultData ) )
			else:
				bareLabel = label.replace( "RGBA", "" ).replace( "RGB", "" )
				channelName = bareLabel
				if label.startswith( "custom" ):
					if channelName in usedNames:
						suffix = 2
						while True:
							channelName = bareLabel + str( suffix )
							if not channelName in usedNames:
								break
							suffix += 1
					categories["Custom"].append( ( label, channelName, defaultData ) )
				else:
					if channelName in usedNames:
						continue
					categories["Standard"].append( ( label, channelName, defaultData ) )


		for category in [ "Standard", "Custom", "Advanced" ]:
			for ( menuLabel, channelName, defaultData ) in categories[category]:
				result.append(
					"/" + category + "/" + menuLabel,
					{
						"command" : functools.partial( Gaffer.WeakMethod( self.__addPlug ), channelName, defaultData ),
					}
				)

		return result

	def __addPlug( self, name, defaultData ) :

		alphaValue = None

		if isinstance( defaultData, IECore.Color4fData ):
			alphaValue = Gaffer.FloatPlug( "value", Gaffer.Plug.Direction.In, defaultData.value.a )
			defaultData = IECore.Color3fData( imath.Color3f( defaultData.value.r, defaultData.value.g, defaultData.value.b ) )

		if defaultData == None:
			plugName = "closure"
			name = ""
			valuePlug = GafferOSL.ClosurePlug( "value" )
		else:
			plugName = "channel"
			valuePlug = Gaffer.PlugAlgo.createPlugFromData( "value", Gaffer.Plug.Direction.In, Gaffer.Plug.Flags.Default, defaultData )

		with Gaffer.UndoScope( self.getPlug().ancestor( Gaffer.ScriptNode ) ) :
			self.getPlug().addChild( Gaffer.NameValuePlug( name, valuePlug, True, plugName, Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
			if alphaValue:
				self.getPlug().addChild(
					Gaffer.NameValuePlug( name + ".A" if name else "A", alphaValue, True, plugName, Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic )
				)

def __channelLabelFromPlug( plug ):
	if plug.typeId() == GafferOSL.ClosurePlug.staticTypeId():
		return plug.parent().getName()
	elif plug.typeId() == Gaffer.Color3fPlug.staticTypeId() and plug.parent()["name"].getValue() == "":
		return "[RGB]"
	else:
		return plug.parent()["name"].getValue()

##########################################################################
# Metadata
##########################################################################

Gaffer.Metadata.registerNode(

	GafferOSL.OSLImage,

	"description",
	"""
	Executes OSL shaders to perform image processing. Use the shaders from
	the OSL/ImageProcessing menu to read values from the input image and
	then write values back to it.
	""",

	"plugAdderOptions", IECore.CompoundData( _channelNamesOptions ),

	"layout:activator:defaultFormatActive", lambda node : not node["in"].getInput(),

	plugs = {
		"defaultFormat" : [
			"description",
			"""
			The resolution and aspect ratio to output when there is no input image provided.
			""",
			"layout:activator", "defaultFormatActive",
		],
		"channels" : [
			"description",
			"""
			Define image channels to output by adding child plugs and connecting
			corresponding OSL shaders.  You can drive RGB layers with a color,
			or connect individual channels to a float.

			If you want to add multiple channels at once, you can also add a closure plug,
			which can accept a connection from an OSLCode with a combined output closure.
 			""",
			"layout:customWidget:footer:widgetType", "GafferOSLUI.OSLImageUI._ChannelsFooter",
			"layout:customWidget:footer:index", -1,
			"nodule:type", "GafferUI::CompoundNodule",
			"noduleLayout:section", "left",
			"noduleLayout:spacing", 0.2,
			"plugValueWidget:type", "GafferUI.LayoutPlugValueWidget",

			# Add + button for showing and hiding parameters in the GraphEditor
			"noduleLayout:customGadget:addButton:gadgetType", "GafferOSLUI.OSLImageUI.PlugAdder",
		],
		"channels.*" : [

			"deletable", True,
			# Although the parameters plug is positioned
			# as we want above, we must also register
			# appropriate values for each individual parameter,
			# for the case where they get promoted to a box
			# individually.
			"noduleLayout:section", "left",
			"nodule:type", "GafferUI::CompoundNodule",
			"nameValuePlugPlugValueWidget:ignoreNamePlug", lambda plug : isinstance( plug["value"], GafferOSL.ClosurePlug ),
		],
		"channels.*.name" : [
			"nodule:type", "",
			"stringPlugValueWidget:placeholderText", lambda plug : "[RGB]" if isinstance( plug.parent()["value"], Gaffer.Color3fPlug ) else None,
		],
		"channels.*.enabled" : [
			"nodule:type", "",
		],
		"channels.*.value" : [

			# Although the parameters plug is positioned
			# as we want above, we must also register
			# appropriate values for each individual parameter,
			# for the case where they get promoted to a box
			# individually.
			"noduleLayout:section", "left",
			"noduleLayout:label", __channelLabelFromPlug,
			"ui:visibleDimensions", lambda plug : 2 if hasattr( plug, "interpretation" ) and plug.interpretation() == IECore.GeometricData.Interpretation.UV else None,
		],
	}

)
